runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -0,0 +1,180 @@
+     */
+    private static final String RESTORE_TEMP_DIR_PREFIX = "corfu_restore_";
+
+    /***
+     * @param filePath      - the path of backup tar file
+     * @param runtime       - the runtime which is performing the restore
+     * @throws IOException when failed to create the temp directory
+     */
+    public Restore(String filePath, CorfuRuntime runtime) throws IOException {
+        this.filePath = filePath;
+        this.restoreTempDirPath = Files.createTempDirectory(RESTORE_TEMP_DIR_PREFIX).toString();
+        this.streamIDs = new ArrayList<>();
+        this.corfuStore = new CorfuStore(runtime);
+    }
+
+    public boolean start() throws IOException {
+        File backupTarFile = new File(filePath);
+        if (!backupTarFile.exists()) {
+            return false;
+        }
+
+        openTarFile();
+
+        if (!verify()) {
+            return false;
+        }
+
+        if (!restore()) {
+            cleanup();
+            return false;
+        }
+
+        cleanup();
+        return true;
+    }
+
+    private boolean restore() throws IOException {
+        for (UUID streamId : streamIDs) {
+            String fileName = restoreTempDirPath + File.separator + streamId;
+            if (!restoreTable(fileName, streamId)) {
+                log.error("Failed to restore table {} from temp file {}", streamId, fileName);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Restore a single table
+     * @param fileName   - the name of the temp backup file
+     * @param streamId   - the stream ID of the table which is to be restored
+     * @return true if restore succeeded
+     * @throws IOException when 1) backup file is not found or 2) failed to read file
+     */
+    private boolean restoreTable(String fileName, UUID streamId) throws IOException {
+        FileInputStream fileInput = null;
+        try {
+            fileInput = new FileInputStream(fileName);
+        } catch (FileNotFoundException e) {
+            log.error("restoreTable can not find file {}", fileName);
+            return false;
+        }
+        long numEntries = 0;
+
+        // Clear table before restore
+        TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE);
+        SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+        txn.logUpdate(streamId, entry);
+
+        // For each opaque entry, write a transaction to the database.
+        try {
+            while (fileInput.available()> 0) {
+                OpaqueEntry opaqueEntry = OpaqueEntry.read(fileInput);
+                List<SMREntry> smrEntries = opaqueEntry.getEntries().get(streamId);
+                if (smrEntries == null || smrEntries.isEmpty()) {
+                    continue;
+                }
+
+                txn.logUpdate(streamId, smrEntries);
+                numEntries++;
+                log.debug("write uuid {} with {} numEntries", streamId, numEntries);
```

Comment:
Should this line be outside of the while loop?

Reply:
Done.